### ISSUE TITLE
[Feature] Fault permeability specification

### ISSUE THREAD

**Issue by joshua-white** [GEOSX/GEOSX#1124]
**What is the requested feature?**
Faults can now be identified as node sets.  We would like to enable the ability to treat faults as either flow conduits or flow barriers.  There are two cases to consider:
* Flow barrier: we just need to cut the transmissibility graph across faces connected through a fault face
* Flow conduit: we need to add a FaceElement to the fault face, and be able to specify across fault and along fault perms

We had some good discussion on this at the last dev mtg. @rrsettgast and @klevzoff, could you summarize your design suggestions?

**Comment by ValarcherL** [comment]
I think the easiest would be to keep the actual design (Edge and Vertex) adding the possibility for Vertex to represent not only element but also faces, and maybe even in the future other elements like nodes (to include the connection with wells). The edges, representing the flow, support the transmissibility, by setting it to 0 on both edges between the face and the element we can represent the first case. For the second it would imply to add face to face connections supporting the permeability along fault.
I think there is other solutions so if you think of something else, that may be more achievable in GEOSX.

**Comment by joshua-white** [comment]
Thanks @ValarcherL.  Your comments remind me we should be clear about terminology, as we are mixing two categories:

Graph objects (stencil-based): vertex and edge
Geometric objects (mesh-based): element, face, node, edge, wells

I propose we try to always refer to a "graph vertex" or "graph edge" for graph objects, and things like CellElement, FaceElement, WellElement, etc. wherever possible when referring to our geometric objects.  There is bound to be some ambiguity though.

**Comment by klevzoff** [comment]
My proposal for a simple, temporary solution was:
 1. The mesh generator optionally allocates on `FaceManager` a real-valued field named `transMultiplier` (or something)
 1. If the field is present, `TwoPointFluxApproximation` object applies that coefficient to computed transmissibility across the face.
 1. If the value is set to zero, the stencil entry is effectively discarded. We could decide whether it's better to keep the entry with zero weights, or not keep it at all.

This has advantages:
 * simple and fast to implement
 * does not rely on explicit surface generation, everything works just on mesh faces
 * flexible for engineering workflows since multiplier values could be set to anything (not just zero)
 * the multiplier field could also be set via `FieldSpecification` mechanism, so one could use it with internal mesh generator and geometry objects to create simple sealing faults; in other words, no need for sophisticated mesh import

and disadvantages:
 * only cross-fault flow is affected and flow along the fault is not considered, so no way to properly model leaking faults
 * setting arbitrary coefficients is not based on physical principles

(This proposal is based on existing implementation does not consider @ValarcherL 's work on graph structures, but I think those are sort of parallel to each other)

**Comment by AntoineMazuyer** [comment]
@joshua-white Thanks for the issue

@klevzoff some comments : 
 >The mesh generator optionally allocates on FaceManager a real-valued field named transMultiplier (or something) 

Agree, but I think that the faces are not built during the `GenerateMesh` process, but after. When we load the mesh, (or build it using the `InternalMeshGenerator`, only pointsets composed of vertices for each faults or fractures are known.

 > the multiplier field could also be set via FieldSpecification mechanism, so one could use it with internal mesh generator and geometry-based face sets to create simple sealing faults; in other words, no need for sophisticated mesh import.

Can we refer to a specific "face set" using the `FieldSpecification` mechanism ? (I guess the answer is yes)

**Comment by klevzoff** [comment]
> Agree, but I think that the faces are not built during the `GenerateMesh` process, but after. When we load the mesh, (or build it using the `InternalMeshGenerator`, only pointsets composed of vertices for each faults or fractures are known.

Hmm. In that case, I'd say we stick to just importing surfaces from the mesh generator, and later set the trans coefficient via `FieldSpecification` only. A more general solution would be to split mesh import and property import into separate routines.

> Can we refer to a specific "face set" using the `FieldSpecification` mechanism ? (I guess the answer is yes)

Indeed, face sets are generated based on node sets and have the same names. `SinglePhaseFlow` uses this capability to implement face-based BC.
